specifyweb/frontend/js_src/lib/components/Notifications/hooks.tsx

```diff
@@ -10,6 +10,7 @@ import type { GenericNotification } from './NotificationRenderers';
 
 const INITIAL_INTERVAL = 5000;
 const INTERVAL_MULTIPLIER = 1.1;
+let addSinceToNotification = true;
```

Comment:
this is a wrong location. it should be as a ref inside of `useNotificationsFetch()`
the notifications state is initialized to undefined when the hook is first called - so this boolean flag should be initialized at the same time

the Notifications component may be re-rendered completely (maybe when user changes the menu position)

Reply:
Explanation on why this could lead to a bug, and should be changed to a ref:

This value is used to denote whether it's a first fetch of notifications (thus need to fetch all) or if we are just appending to what we have fetched already (just fetch only new).

But, the notifications themself, are stored in a state in the hook - imagine this hook is instantiated again (meaning rendered from scratch, and state values are not preserved) (will mention why that may happen later, just assume it does for now), then notifications state is initialized from scratch with the undefined value.

yet addSinceToNotification would be false, because you set it to false on the useEffect execution in the previous instance of this hook. Now, you have a bug - no previous notifications will be fetched.

The solution is to store the "addSinceToNotification" value in the same place as where the notifications state is, so that if the hook is instantiated again, the notifications state is reset to undefined AND addSinceToNotification is reset to true at the same time.

That is why I suggested using a ref for "addSinceToNotification" - you don't care about this value for rendering UI components (unlike "notifications" state), and you don't care about being notified when this value changes - all you care about is having the newest value of this variable when you want check it's value in doFetch. That is why useRef is perfect for it, rather than useState.

Does this make sense?

Now for the reasons why hook may be instantiated again:
- Imagine some other code elsewhere wants to use this hook in another place. Possible reasons:
  - maybe you want an admin panel that lists notifications per user?
  - or the ability to see notifications in user preferences (so that you can customize them and see impact right away)
  - or someone wants to reuse this logic for something totally different (just like @realVinayak did with [drag and drop hook](https://github.com/specify/specify7/blob/298f7ca5d7d709e3a9b9ad6813e17e9a237b5c08/specifyweb/frontend/js_src/lib/components/Molecules/useDragDropFiles.ts) - this logic was intended originally to only be used in the FilePicker, but now the hook is used outside of FilePicker too). By not assuming that there is only one instance of this code executing at one time, reusing code becomes easy, without having to double check all logic to make sure no unsafe assumptions were made)
- The entire header component may be recreated. Possible cases:
  - User changes header position in user preferences
  - Or an error happens, which un-renders the entire app, but then we implement error recovery that leads user back to the home page
- Probably other cases too. But the point is it's unsafe, and confusing to those reading this code

---


looking at this again, you might be able to reduce the need for this variable by just checking inside of useEffect if notifications are undefined (without adding notifications to the dependency array) - not sure if that would work, but if it would that would be neat to try